pkg/route/hostassignment/externalcertificate.go

```diff
@@ -0,0 +1,34 @@
+package hostassignment
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	routev1 "github.com/openshift/api/route/v1"
+	routecommon "github.com/openshift/library-go/pkg/route"
+)
+
+// validateHostExternalCertificate checks if the user has permissions to create and update
+// custom-host subresource of routes. This check is required to be done prior to ValidateHostUpdate()
+// since updating hosts while using externalCertificate is contingent on the user having both these
+// permissions. The ValidateHostUpdate() cannot differentiate if the certificate has changed since
+// now the certificates will be present as a secret object, due to this it proceeds with the assumption
+// that the certificate has changed when the route has externalCertificate set.
+func ValidateHostExternalCertificate(ctx context.Context, new, older *routev1.Route, sarc routecommon.SubjectAccessReviewCreator, opts routecommon.RouteValidationOptions) field.ErrorList {
+	newTLS := new.Spec.TLS
+	oldTLS := older.Spec.TLS
+
+	if !opts.AllowExternalCertificates {
```

Comment:
I think `ValidateHostExternalCertificate` is the [only place](https://github.com/openshift/openshift-apiserver/pull/407/files#diff-3c95009cb726729b48ee6ff52d62e70f3352aa6af1a0dedd2047ecdc5a398ce9R125-R127)  where apiserver checks for featuregate, rest places it relies on `RouteValidationOptions` struct to decide inside library go. 

I see challenges in removing this:

- `AllocateHost` and `ValidateHostUpdate` also rely on this struct and take decisions accordingly, how are we going to handle logic for them?
- If we want to add new methods for route validation, then I think we will need to have 2 methods, for example `ValidateRouteExternalCertificate` and `ValidateRouteUpdateExternalCertificate` in library-go as well as in [api-server's validation.go](https://github.com/openshift/openshift-apiserver/blob/master/pkg/route/apis/route/validation/validation.go). I think it can cause more overhead in longer run, and prone to future refactoring. 

IMHO, we should rather remove the feature check in api-server and let the flow decide based upon `RouteValidationOptions` here in library-go to remove redundancy. (all the functions are already doing that).

Reply:
I didn't get it properly. We should keep `opts.AllowExternalCertificates` here in library-go and populate this from api-server.  
Api-server will not decide the flow; rather it will rely on this field to decide the flow in library-go. 

This will give more control and less code refactoring in the future.

Reply:
I agree on letting library-go decide on the flow. What I'm wondering about is why there is a dedicated method for the external certificate, why this code is not integrated into `ValidateHostUpdate`.

Reply:
I think it had been touched upon during EP review, and also documented [here](https://github.com/thejasn/enhancements/blob/d4c7701193646105e2c60b2b996047ddfed083a4/enhancements/ingress/route-secret-injection-for-external-certificate-management.md#implementation-detailsnotesconstraints-optional). Let me know if that has changed afterwards.